% Generated by Sphinx.
\def\sphinxdocclass{report}
\documentclass[letterpaper,10pt,english]{sphinxmanual}
\usepackage[utf8]{inputenc}
\DeclareUnicodeCharacter{00A0}{\nobreakspace}
\usepackage[T1]{fontenc}
\usepackage{babel}
\usepackage{times}
\usepackage[Bjarne]{fncychap}
\usepackage{longtable}
\usepackage{sphinx}
\usepackage{multirow}


\title{gFunc Documentation}
\date{January 11, 2013}
\release{0.1}
\author{Augustine (Gus) Dunn}
\newcommand{\sphinxlogo}{}
\renewcommand{\releasename}{Release}
\makeindex

\makeatletter
\def\PYG@reset{\let\PYG@it=\relax \let\PYG@bf=\relax%
    \let\PYG@ul=\relax \let\PYG@tc=\relax%
    \let\PYG@bc=\relax \let\PYG@ff=\relax}
\def\PYG@tok#1{\csname PYG@tok@#1\endcsname}
\def\PYG@toks#1+{\ifx\relax#1\empty\else%
    \PYG@tok{#1}\expandafter\PYG@toks\fi}
\def\PYG@do#1{\PYG@bc{\PYG@tc{\PYG@ul{%
    \PYG@it{\PYG@bf{\PYG@ff{#1}}}}}}}
\def\PYG#1#2{\PYG@reset\PYG@toks#1+\relax+\PYG@do{#2}}

\expandafter\def\csname PYG@tok@gd\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.63,0.00,0.00}{##1}}}
\expandafter\def\csname PYG@tok@gu\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.50,0.00,0.50}{##1}}}
\expandafter\def\csname PYG@tok@gt\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.25,0.82}{##1}}}
\expandafter\def\csname PYG@tok@gs\endcsname{\let\PYG@bf=\textbf}
\expandafter\def\csname PYG@tok@gr\endcsname{\def\PYG@tc##1{\textcolor[rgb]{1.00,0.00,0.00}{##1}}}
\expandafter\def\csname PYG@tok@cm\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}}
\expandafter\def\csname PYG@tok@vg\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.38,0.84}{##1}}}
\expandafter\def\csname PYG@tok@m\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@mh\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@cs\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}\def\PYG@bc##1{\setlength{\fboxsep}{0pt}\colorbox[rgb]{1.00,0.94,0.94}{\strut ##1}}}
\expandafter\def\csname PYG@tok@ge\endcsname{\let\PYG@it=\textit}
\expandafter\def\csname PYG@tok@vc\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.38,0.84}{##1}}}
\expandafter\def\csname PYG@tok@il\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@go\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.19,0.19,0.19}{##1}}}
\expandafter\def\csname PYG@tok@cp\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@gi\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.63,0.00}{##1}}}
\expandafter\def\csname PYG@tok@gh\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.00,0.50}{##1}}}
\expandafter\def\csname PYG@tok@ni\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.84,0.33,0.22}{##1}}}
\expandafter\def\csname PYG@tok@nl\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.13,0.44}{##1}}}
\expandafter\def\csname PYG@tok@nn\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.05,0.52,0.71}{##1}}}
\expandafter\def\csname PYG@tok@no\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.38,0.68,0.84}{##1}}}
\expandafter\def\csname PYG@tok@na\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@nb\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@nc\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.05,0.52,0.71}{##1}}}
\expandafter\def\csname PYG@tok@nd\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.33,0.33,0.33}{##1}}}
\expandafter\def\csname PYG@tok@ne\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@nf\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.02,0.16,0.49}{##1}}}
\expandafter\def\csname PYG@tok@si\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.44,0.63,0.82}{##1}}}
\expandafter\def\csname PYG@tok@s2\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@vi\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.38,0.84}{##1}}}
\expandafter\def\csname PYG@tok@nt\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.02,0.16,0.45}{##1}}}
\expandafter\def\csname PYG@tok@nv\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.38,0.84}{##1}}}
\expandafter\def\csname PYG@tok@s1\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@gp\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.78,0.36,0.04}{##1}}}
\expandafter\def\csname PYG@tok@sh\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@ow\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@sx\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.78,0.36,0.04}{##1}}}
\expandafter\def\csname PYG@tok@bp\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@c1\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}}
\expandafter\def\csname PYG@tok@kc\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@c\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}}
\expandafter\def\csname PYG@tok@mf\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@err\endcsname{\def\PYG@bc##1{\setlength{\fboxsep}{0pt}\fcolorbox[rgb]{1.00,0.00,0.00}{1,1,1}{\strut ##1}}}
\expandafter\def\csname PYG@tok@kd\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@ss\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.32,0.47,0.09}{##1}}}
\expandafter\def\csname PYG@tok@sr\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.14,0.33,0.53}{##1}}}
\expandafter\def\csname PYG@tok@mo\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@mi\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@kn\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@o\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PYG@tok@kr\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@s\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@kp\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@w\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.73,0.73}{##1}}}
\expandafter\def\csname PYG@tok@kt\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.56,0.13,0.00}{##1}}}
\expandafter\def\csname PYG@tok@sc\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@sb\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@k\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@se\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@sd\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}

\def\PYGZbs{\char`\\}
\def\PYGZus{\char`\_}
\def\PYGZob{\char`\{}
\def\PYGZcb{\char`\}}
\def\PYGZca{\char`\^}
\def\PYGZam{\char`\&}
\def\PYGZlt{\char`\<}
\def\PYGZgt{\char`\>}
\def\PYGZsh{\char`\#}
\def\PYGZpc{\char`\%}
\def\PYGZdl{\char`\$}
\def\PYGZti{\char`\~}
% for compatibility with earlier versions
\def\PYGZat{@}
\def\PYGZlb{[}
\def\PYGZrb{]}
\makeatother

\begin{document}

\maketitle
\tableofcontents
\phantomsection\label{index::doc}


Contents:


\chapter{gFunc Tutorial}
\label{tutorial:welcome-to-gfunc-s-documentation}\label{tutorial::doc}\label{tutorial:gfunc-tutorial}
Here be the learing of the majic.


\chapter{Project Summary}
\label{project::doc}\label{project:project-summary}

\section{Goals}
\label{project:goals}\begin{itemize}
\item {} 
Galaxy integration

\item {} 
Graph based integration of multiple OMICs data streams

\item {} 
Rank/cull gene names based on multidimentional association with user defined weigh structure.

\end{itemize}


\section{Lessons Learned}
\label{project:lessons-learned}\begin{itemize}
\item {} 
None so far!

\end{itemize}


\chapter{gFunc Auto-Generated Code Documentation}
\label{code:gfunc-auto-generated-code-documentation}\label{code:module-gfunc.__init__}\label{code::doc}\index{gfunc.\_\_init\_\_ (module)}

\section{gfunc}
\label{code:gfunc}
Root directory and main package namespace.
\phantomsection\label{code:module-gfunc.parsers.__init__}\index{gfunc.parsers.\_\_init\_\_ (module)}

\section{parsers}
\label{code:parsers}
Code supporting parsing of supported raw data files.
\phantomsection\label{code:module-gfunc.parsers.base}\index{gfunc.parsers.base (module)}

\subsection{base.py}
\label{code:base-py}
Code defining base parser class: TODO.
\index{GFuncParserBase (class in gfunc.parsers.base)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.parsers.base.GFuncParserBase}\pysigline{\strong{class }\code{gfunc.parsers.base.}\bfcode{GFuncParserBase}}
TODO: define once I have settled on common needs.

\end{fulllineitems}

\phantomsection\label{code:module-gfunc.parsers.Cufflinks}\index{gfunc.parsers.Cufflinks (module)}

\subsection{Cufflinks.py}
\label{code:cufflinks-py}
Code supporting parsing of Cufflinks type raw data files.
\index{CDiffFpkmTrackerParser (class in gfunc.parsers.Cufflinks)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.parsers.Cufflinks.CDiffFpkmTrackerParser}\pysiglinewithargsret{\strong{class }\code{gfunc.parsers.Cufflinks.}\bfcode{CDiffFpkmTrackerParser}}{\emph{cuffdiff\_fpkm\_path}, \emph{species}, \emph{cuffdiff\_exp\_path=None}, \emph{name\_col='nearest\_ref\_id'}, \emph{combine\_transcripts=True}, \emph{tx\_2\_gene=None}}{}
Class to accept CuffDiff FKPM data table and init/update the relevant gFuncNode Objects.
\index{\_\_init\_\_() (gfunc.parsers.Cufflinks.CDiffFpkmTrackerParser method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.parsers.Cufflinks.CDiffFpkmTrackerParser.__init__}\pysiglinewithargsret{\bfcode{\_\_init\_\_}}{\emph{cuffdiff\_fpkm\_path}, \emph{species}, \emph{cuffdiff\_exp\_path=None}, \emph{name\_col='nearest\_ref\_id'}, \emph{combine\_transcripts=True}, \emph{tx\_2\_gene=None}}{}
Test doc for init'ing CuffDiff parser.

\end{fulllineitems}

\index{resgister\_nodes\_and\_edges() (gfunc.parsers.Cufflinks.CDiffFpkmTrackerParser method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.parsers.Cufflinks.CDiffFpkmTrackerParser.resgister_nodes_and_edges}\pysiglinewithargsret{\bfcode{resgister\_nodes\_and\_edges}}{\emph{node\_dict}, \emph{edge\_dict}, \emph{graph}}{}
Parses each row from the CuffDiff FKPM data table and either adds the data to the relevant
GFuncNode in node\_dict or creates one and adds it to that then registers it in node\_dict.

\end{fulllineitems}


\end{fulllineitems}

\index{am\_i\_sigDiff() (in module gfunc.parsers.Cufflinks)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.parsers.Cufflinks.am_i_sigDiff}\pysiglinewithargsret{\code{gfunc.parsers.Cufflinks.}\bfcode{am\_i\_sigDiff}}{\emph{xloc\_number}, \emph{expDiffTable\_dict}, \emph{q\_thresh}}{}~
\begin{DUlineblock}{0em}
\item[] for lines in cuffdiff\_fpkm\_table:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] Return \code{True} if:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] at least one of current line's XLOC\_xxxx pair tests in \code{expDiffTable\_dict} has \emph{q\_val} \textless{}= \code{q\_thresh}
\end{DUlineblock}
\item[] Else Return \code{False}
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}

\index{build\_expDiffTable\_dict() (in module gfunc.parsers.Cufflinks)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.parsers.Cufflinks.build_expDiffTable_dict}\pysiglinewithargsret{\code{gfunc.parsers.Cufflinks.}\bfcode{build\_expDiffTable\_dict}}{\emph{expDiffTable\_path}}{}
Build isoformExpDiffTable\_dict:
\begin{description}
\item[{Keys:}] \leavevmode
XLOC\_xxxxx

\item[{Values:}] \leavevmode
namedtuple-ified rows with same XLOC\_xxxxx

\end{description}

\end{fulllineitems}

\index{transfer\_nearestRefgeneSymbol\_from\_isoform\_to\_gene\_tracking() (in module gfunc.parsers.Cufflinks)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.parsers.Cufflinks.transfer_nearestRefgeneSymbol_from_isoform_to_gene_tracking}\pysiglinewithargsret{\code{gfunc.parsers.Cufflinks.}\bfcode{transfer\_nearestRefgeneSymbol\_from\_isoform\_to\_gene\_tracking}}{\emph{isoform\_fpkm\_path}, \emph{gene\_fpkm\_path}}{}
TODO: doc

\end{fulllineitems}

\phantomsection\label{code:module-gfunc.parsers.edge_lists}\index{gfunc.parsers.edge\_lists (module)}

\subsection{edge\_lists.py}
\label{code:edge-lists-py}
Code supporting parsing of lists into edge connections.
\index{OneToOneOrthoListParser (class in gfunc.parsers.edge\_lists)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.parsers.edge_lists.OneToOneOrthoListParser}\pysiglinewithargsret{\strong{class }\code{gfunc.parsers.edge\_lists.}\bfcode{OneToOneOrthoListParser}}{\emph{list\_path='`}, \emph{divergence\_info=None}, \emph{relation\_type='one\_to\_one\_ortholog'}}{}
Class to accept list of rows when each item/node\_name in the row should have edges to all other items in the row
and init the relevant gFuncNode/gFuncEdges Objects.  Each Column should have a header that is supported
by gfunc.fileIO.tableFile2namedTuple representing the species of the nodeName in that column (Anopheles\_gambiae).
\index{\_\_init\_\_() (gfunc.parsers.edge\_lists.OneToOneOrthoListParser method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.parsers.edge_lists.OneToOneOrthoListParser.__init__}\pysiglinewithargsret{\bfcode{\_\_init\_\_}}{\emph{list\_path='`}, \emph{divergence\_info=None}, \emph{relation\_type='one\_to\_one\_ortholog'}}{}
TODO: doc for init'ing.

\end{fulllineitems}

\index{resgister\_nodes\_and\_edges() (gfunc.parsers.edge\_lists.OneToOneOrthoListParser method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.parsers.edge_lists.OneToOneOrthoListParser.resgister_nodes_and_edges}\pysiglinewithargsret{\bfcode{resgister\_nodes\_and\_edges}}{\emph{node\_dict}, \emph{edge\_dict}, \emph{graph}}{}
Iterates through every row in list\_path ensuring that
a GFuncNode exists for each nodeName and is registered.  GFuncNodes are
initialized with basic info (name,species) if it doesnt already exist.
Then GFuncEdge objects are registered/initialized for nodeName combinations
in each row while setting \textless{}relation\_type\textgreater{} data\_type to `True' for each edge.

\end{fulllineitems}


\end{fulllineitems}

\index{combine\_multiple\_one2one\_tables() (in module gfunc.parsers.edge\_lists)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.parsers.edge_lists.combine_multiple_one2one_tables}\pysiglinewithargsret{\code{gfunc.parsers.edge\_lists.}\bfcode{combine\_multiple\_one2one\_tables}}{\emph{path\_list, species\_prefixes={[}'AGAP', `AAEL', `CPIJ'{]}}}{}
Builds a set of linked dicts with:
\begin{description}
\item[{Keys:}] \leavevmode
GeneName

\item[{Values:}] \leavevmode
Link to other GeneName keys that are supossed one2one orthologs

\end{description}

\end{fulllineitems}

\index{follow\_all\_links() (in module gfunc.parsers.edge\_lists)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.parsers.edge_lists.follow_all_links}\pysiglinewithargsret{\code{gfunc.parsers.edge\_lists.}\bfcode{follow\_all\_links}}{\emph{graph}, \emph{node}}{}
Return all nodes in connected subgraph wrt node.

\end{fulllineitems}

\phantomsection\label{code:module-gfunc.parsers.ETE}\index{gfunc.parsers.ETE (module)}

\subsection{ETE.py}
\label{code:ete-py}
Code supporting parsing phyloXML files using the ETE2 package.
\index{PhyloXMLParser (class in gfunc.parsers.ETE)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.parsers.ETE.PhyloXMLParser}\pysiglinewithargsret{\strong{class }\code{gfunc.parsers.ETE.}\bfcode{PhyloXMLParser}}{\emph{phyloXML\_path='`}, \emph{species=}\optional{}, \emph{pickle\_path=None}}{}
Class to accept PhyloXML files or directories and init the relevant gFuncNode/gFuncEdges Objects.

RIGHT NOW: only used for branch\_length
\index{\_\_init\_\_() (gfunc.parsers.ETE.PhyloXMLParser method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.parsers.ETE.PhyloXMLParser.__init__}\pysiglinewithargsret{\bfcode{\_\_init\_\_}}{\emph{phyloXML\_path='`}, \emph{species=}\optional{}, \emph{pickle\_path=None}}{}
Test doc for init'ing.

\end{fulllineitems}

\index{get\_distance() (gfunc.parsers.ETE.PhyloXMLParser method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.parsers.ETE.PhyloXMLParser.get_distance}\pysiglinewithargsret{\bfcode{get\_distance}}{\emph{leaf1}, \emph{leaf2}}{}
For two leaf objs in a common tree, returns the branch length that
separates them.

\end{fulllineitems}

\index{get\_species() (gfunc.parsers.ETE.PhyloXMLParser method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.parsers.ETE.PhyloXMLParser.get_species}\pysiglinewithargsret{\bfcode{get\_species}}{\emph{leaf}}{}
Returns the species scientific name of a leaf obj.

\end{fulllineitems}

\index{resgister\_nodes\_and\_edges() (gfunc.parsers.ETE.PhyloXMLParser method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.parsers.ETE.PhyloXMLParser.resgister_nodes_and_edges}\pysiglinewithargsret{\bfcode{resgister\_nodes\_and\_edges}}{\emph{node\_dict}, \emph{edge\_dict}, \emph{graph}}{}
Iterates through every leaf in every tree in self.trees ensuring that
a GFuncNode exists for each leaf and is registered.  GFuncNodes are
initialized with basic info (name,species) if it doesnt already exist.
Then GFuncEdge objects are registered/initialized for leaf combinations
in each tree while setting `branch\_length' data for each edge.

\end{fulllineitems}


\end{fulllineitems}

\index{load\_phyloXMLs() (in module gfunc.parsers.ETE)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.parsers.ETE.load_phyloXMLs}\pysiglinewithargsret{\code{gfunc.parsers.ETE.}\bfcode{load\_phyloXMLs}}{\emph{path}, \emph{species=None}, \emph{pickle\_path=None}}{}
Loads at least one phyloXML file and returns a Phyloxml() project containing at least one phyloXML tree.
If path is a directory, all subdirectories are scrubed for xml files too.

If species!=None, prunes trees to leaves that are of the supplied scientific species names (can save a LOT
of memory if the number of trees is large).

path = str()
species = list()

\end{fulllineitems}

\index{prune\_trees\_by\_species() (in module gfunc.parsers.ETE)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.parsers.ETE.prune_trees_by_species}\pysiglinewithargsret{\code{gfunc.parsers.ETE.}\bfcode{prune\_trees\_by\_species}}{\emph{ete2\_tree\_list}, \emph{species\_list}}{}
XXXX

\end{fulllineitems}

\phantomsection\label{code:module-gfunc.parsers.GTF}\index{gfunc.parsers.GTF (module)}

\subsection{GTF.py}
\label{code:gtf-py}
Code supporting parsing/indexing of GTF/GFF data files.

NOTE: The ``business'' end of this code uses (read: depends on) the gtf\_to\_genes module:
\begin{quote}

Metadata-Version: 1.0
Name: gtf-to-genes
Version: 1.07
Summary: Fast GTF parser
Home-page: \href{http://code.google.com/p/gtf-to-genes/}{http://code.google.com/p/gtf-to-genes/}
Author: Leo Goodstadt
Author-email: \href{mailto:gtf\_to\_genes@llew.org.uk}{gtf\_to\_genes@llew.org.uk}
License: MIT
\end{quote}
\phantomsection\label{code:module-gfunc.parsers.JASPAR}\index{gfunc.parsers.JASPAR (module)}

\subsection{JASPAR.py}
\label{code:jaspar-py}
Code supporting reading and writing of basic PSSMs representing TFBS profiles; especially in JASPAR format.
\index{BasicTFBSParser (class in gfunc.parsers.JASPAR)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.parsers.JASPAR.BasicTFBSParser}\pysiglinewithargsret{\strong{class }\code{gfunc.parsers.JASPAR.}\bfcode{BasicTFBSParser}}{\emph{tfbs\_path}}{}
Class to accept TFBS profile data table from `find\_motifs.py' and init/update the relevant gFuncNode Objects.
\index{\_\_init\_\_() (gfunc.parsers.JASPAR.BasicTFBSParser method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.parsers.JASPAR.BasicTFBSParser.__init__}\pysiglinewithargsret{\bfcode{\_\_init\_\_}}{\emph{tfbs\_path}}{}
Test doc for init'ing TFBS parser.

\end{fulllineitems}

\index{resgister\_nodes\_and\_edges() (gfunc.parsers.JASPAR.BasicTFBSParser method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.parsers.JASPAR.BasicTFBSParser.resgister_nodes_and_edges}\pysiglinewithargsret{\bfcode{resgister\_nodes\_and\_edges}}{\emph{node\_dict}, \emph{edge\_dict}, \emph{graph}}{}
Parses each row from the TFBS data table and either adds the data to the relevant
GFuncNode in node\_dict or creates one and adds it to that then registers it in node\_dict.

\end{fulllineitems}


\end{fulllineitems}

\index{ParseJasparMatrixOnly (class in gfunc.parsers.JASPAR)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.parsers.JASPAR.ParseJasparMatrixOnly}\pysiglinewithargsret{\strong{class }\code{gfunc.parsers.JASPAR.}\bfcode{ParseJasparMatrixOnly}}{\emph{filePath}}{}
Returns a record-by-record motif parser for JASPAR matric\_only.txt files analogous to file.readline().

Example:

\begin{Verbatim}[commandchars=\\\{\}]
\textgreater{}MA0001.1 AGL3
A  [ 0  3 79 40 66 48 65 11 65  0 ]
C  [94 75  4  3  1  2  5  2  3  3 ]
G  [ 1  0  3  4  1  0  5  3 28 88 ]
T  [ 2 19 11 50 29 47 22 81  1  6 ]
\textgreater{}MA0002.1 RUNX1
A  [10 12  4  1  2  2  0  0  0  8 13 ]
C  [ 2  2  7  1  0  8  0  0  1  2  2 ]
G  [ 3  1  1  0 23  0 26 26  0  0  4 ]
T  [11 11 14 24  1 16  0  0 25 16  7 ]
\end{Verbatim}
\index{\_\_init\_\_() (gfunc.parsers.JASPAR.ParseJasparMatrixOnly method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.parsers.JASPAR.ParseJasparMatrixOnly.__init__}\pysiglinewithargsret{\bfcode{\_\_init\_\_}}{\emph{filePath}}{}
Returns a record-by-record motif parser analogous to file.readline().
Exmpl: parser.next()
Its ALSO an iterator so ``for rec in parser'' works too!

\end{fulllineitems}

\index{next() (gfunc.parsers.JASPAR.ParseJasparMatrixOnly method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.parsers.JASPAR.ParseJasparMatrixOnly.next}\pysiglinewithargsret{\bfcode{next}}{}{}
Reads in next element, parses, and does minimal verification.
\emph{RETURNS:} tuple: (seqName,seqStr)

\end{fulllineitems}

\index{to\_dict() (gfunc.parsers.JASPAR.ParseJasparMatrixOnly method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.parsers.JASPAR.ParseJasparMatrixOnly.to_dict}\pysiglinewithargsret{\bfcode{to\_dict}}{}{}
Returns a single OrderedDict populated with the motifRecs
contained in self.\_file.

\end{fulllineitems}


\end{fulllineitems}

\phantomsection\label{code:module-gfunc.parsers.MAST}\index{gfunc.parsers.MAST (module)}

\subsection{MAST.py}
\label{code:mast-py}
Code supporting parsing of MAST type raw data files.
\phantomsection\label{code:module-gfunc.analysis_classes}\index{gfunc.analysis\_classes (module)}

\section{analysis\_classes.py}
\label{code:analysis-classes-py}
Code defining controller classes for supported analysis types.
\index{BranchLength (class in gfunc.analysis\_classes)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.analysis_classes.BranchLength}\pysiglinewithargsret{\strong{class }\code{gfunc.analysis\_classes.}\bfcode{BranchLength}}{\emph{poll\_me=False}}{}
TODO: doc
\index{\_\_init\_\_() (gfunc.analysis\_classes.BranchLength method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.analysis_classes.BranchLength.__init__}\pysiglinewithargsret{\bfcode{\_\_init\_\_}}{\emph{poll\_me=False}}{}
TODO: doc

\end{fulllineitems}

\index{measure\_relation() (gfunc.analysis\_classes.BranchLength method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.analysis_classes.BranchLength.measure_relation}\pysiglinewithargsret{\bfcode{measure\_relation}}{\emph{gfunc\_edge}}{}
TODO: doc

\end{fulllineitems}


\end{fulllineitems}

\index{ExpressionSimilarity (class in gfunc.analysis\_classes)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.analysis_classes.ExpressionSimilarity}\pysiglinewithargsret{\strong{class }\code{gfunc.analysis\_classes.}\bfcode{ExpressionSimilarity}}{\emph{poll\_me=False}}{}
TODO: doc
\index{\_\_init\_\_() (gfunc.analysis\_classes.ExpressionSimilarity method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.analysis_classes.ExpressionSimilarity.__init__}\pysiglinewithargsret{\bfcode{\_\_init\_\_}}{\emph{poll\_me=False}}{}
TODO: doc

\end{fulllineitems}


\end{fulllineitems}

\index{Metric (class in gfunc.analysis\_classes)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.analysis_classes.Metric}\pysiglinewithargsret{\strong{class }\code{gfunc.analysis\_classes.}\bfcode{Metric}}{\emph{poll\_me=False}}{}
TODO: doc
\index{\_\_init\_\_() (gfunc.analysis\_classes.Metric method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.analysis_classes.Metric.__init__}\pysiglinewithargsret{\bfcode{\_\_init\_\_}}{\emph{poll\_me=False}}{}
TODO: doc

\end{fulllineitems}

\index{mean() (gfunc.analysis\_classes.Metric method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.analysis_classes.Metric.mean}\pysiglinewithargsret{\bfcode{mean}}{\emph{greater\_than=0}}{}
Returns the mean of the encountered values greater than the value provided.

\end{fulllineitems}

\index{measure\_relation() (gfunc.analysis\_classes.Metric method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.analysis_classes.Metric.measure_relation}\pysiglinewithargsret{\bfcode{measure\_relation}}{\emph{gfunc\_edge}}{}
TODO: doc

\end{fulllineitems}

\index{median() (gfunc.analysis\_classes.Metric method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.analysis_classes.Metric.median}\pysiglinewithargsret{\bfcode{median}}{\emph{greater\_than=0}}{}
Returns the median of the encountered values greater than the value provided.

\end{fulllineitems}


\end{fulllineitems}

\index{PhyloExpnCorrelationIndex (class in gfunc.analysis\_classes)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.analysis_classes.PhyloExpnCorrelationIndex}\pysiglinewithargsret{\strong{class }\code{gfunc.analysis\_classes.}\bfcode{PhyloExpnCorrelationIndex}}{\emph{poll\_me=False}}{}
TODO: doc
\index{\_\_init\_\_() (gfunc.analysis\_classes.PhyloExpnCorrelationIndex method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.analysis_classes.PhyloExpnCorrelationIndex.__init__}\pysiglinewithargsret{\bfcode{\_\_init\_\_}}{\emph{poll\_me=False}}{}
TODO: doc

\end{fulllineitems}


\end{fulllineitems}

\index{RelationsHandler (class in gfunc.analysis\_classes)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.analysis_classes.RelationsHandler}\pysiglinewithargsret{\strong{class }\code{gfunc.analysis\_classes.}\bfcode{RelationsHandler}}{\emph{list\_of\_metrics}}{}
Relations are metrics that characterize how an
edge's 2 connected nodes relate given some type
of relationship (branch length, expression profile
similarity, etc).
\index{\_\_init\_\_() (gfunc.analysis\_classes.RelationsHandler method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.analysis_classes.RelationsHandler.__init__}\pysiglinewithargsret{\bfcode{\_\_init\_\_}}{\emph{list\_of\_metrics}}{}
TODO: Doc

\end{fulllineitems}

\index{get\_vote\_types() (gfunc.analysis\_classes.RelationsHandler method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.analysis_classes.RelationsHandler.get_vote_types}\pysiglinewithargsret{\bfcode{get\_vote\_types}}{}{}
TODO: Doc

\end{fulllineitems}

\index{measure\_relations() (gfunc.analysis\_classes.RelationsHandler method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.analysis_classes.RelationsHandler.measure_relations}\pysiglinewithargsret{\bfcode{measure\_relations}}{\emph{edge\_dict}}{}~
\begin{DUlineblock}{0em}
\item[] For each \emph{gfunc\_edge} in \code{edge\_dict}:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] iterates through \emph{metrics}:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] calculates \& stores result in \emph{gfunc\_edge} and \emph{metric\_handler}
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}


\end{fulllineitems}

\index{TFBSSimilarity (class in gfunc.analysis\_classes)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.analysis_classes.TFBSSimilarity}\pysiglinewithargsret{\strong{class }\code{gfunc.analysis\_classes.}\bfcode{TFBSSimilarity}}{\emph{poll\_me=False}}{}
TODO: doc
\index{\_\_init\_\_() (gfunc.analysis\_classes.TFBSSimilarity method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.analysis_classes.TFBSSimilarity.__init__}\pysiglinewithargsret{\bfcode{\_\_init\_\_}}{\emph{poll\_me=False}}{}
\end{fulllineitems}


\end{fulllineitems}

\index{VoteHandler (class in gfunc.analysis\_classes)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.analysis_classes.VoteHandler}\pysiglinewithargsret{\strong{class }\code{gfunc.analysis\_classes.}\bfcode{VoteHandler}}{\emph{graph}}{}
VoteHandlers determine how well the current GFuncNode's
neighborhood agrees with it regarding each type of relation
being used.  The result is a weighted mean, weighted by
a strength (or trustability metric) between each neighbor
and the current node (exp: branch length, p-value, etc).
If no weight relationship is specified, the result is a
standard mean (weights are equal).
\index{\_\_init\_\_() (gfunc.analysis\_classes.VoteHandler method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.analysis_classes.VoteHandler.__init__}\pysiglinewithargsret{\bfcode{\_\_init\_\_}}{\emph{graph}}{}
TODO: Doc

\end{fulllineitems}

\index{set\_vote\_types() (gfunc.analysis\_classes.VoteHandler method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.analysis_classes.VoteHandler.set_vote_types}\pysiglinewithargsret{\bfcode{set\_vote\_types}}{\emph{vote\_types}, \emph{weight\_by=None}}{}
Recieves and stores as LIST `Metric.relation\_metric' string for
each Metric class that needs a vote taken.

weight\_by: one or none of the gfunc\_edge.data key strings to use to weight
the votes of each node's neighbor edge metrics.
(weight\_by=None results in equal weights)

\end{fulllineitems}

\index{take\_votes() (gfunc.analysis\_classes.VoteHandler method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.analysis_classes.VoteHandler.take_votes}\pysiglinewithargsret{\bfcode{take\_votes}}{\emph{node\_list}, \emph{poll\_func=None}}{}
TODO: Doc

\end{fulllineitems}


\end{fulllineitems}

\phantomsection\label{code:module-gfunc.clustering}\index{gfunc.clustering (module)}

\section{clustering.py}
\label{code:clustering-py}
Code supporting efforts to optimize and automate external clustering libraries for gFunc purposes.
\index{plot\_centers\_and\_points() (in module gfunc.clustering)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.clustering.plot_centers_and_points}\pysiglinewithargsret{\code{gfunc.clustering.}\bfcode{plot\_centers\_and\_points}}{\emph{data}, \emph{clusters}, \emph{means}, \emph{truth=None}}{}
\end{fulllineitems}

\phantomsection\label{code:module-gfunc.data_classes}\index{gfunc.data\_classes (module)}\index{Bunch (class in gfunc.data\_classes)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.data_classes.Bunch}\pysiglinewithargsret{\strong{class }\code{gfunc.data\_classes.}\bfcode{Bunch}}{\emph{*args}, \emph{**kwds}}{}
A dict like class to facilitate setting and access to tree-like data.
\index{\_\_init\_\_() (gfunc.data\_classes.Bunch method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.data_classes.Bunch.__init__}\pysiglinewithargsret{\bfcode{\_\_init\_\_}}{\emph{*args}, \emph{**kwds}}{}
\end{fulllineitems}


\end{fulllineitems}

\index{GFuncEdge (class in gfunc.data\_classes)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.data_classes.GFuncEdge}\pysiglinewithargsret{\strong{class }\code{gfunc.data\_classes.}\bfcode{GFuncEdge}}{\emph{node1}, \emph{node2}}{}
TODO: Doc
\index{\_\_init\_\_() (gfunc.data\_classes.GFuncEdge method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.data_classes.GFuncEdge.__init__}\pysiglinewithargsret{\bfcode{\_\_init\_\_}}{\emph{node1}, \emph{node2}}{}
TODO: Doc

\end{fulllineitems}

\index{set\_data() (gfunc.data\_classes.GFuncEdge method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.data_classes.GFuncEdge.set_data}\pysiglinewithargsret{\bfcode{set\_data}}{\emph{data}, \emph{data\_type}}{}
TODO: Doc

\end{fulllineitems}


\end{fulllineitems}

\index{GFuncNode (class in gfunc.data\_classes)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.data_classes.GFuncNode}\pysiglinewithargsret{\strong{class }\code{gfunc.data\_classes.}\bfcode{GFuncNode}}{\emph{name}, \emph{species}, \emph{graph}, \emph{is\_target=False}, \emph{debug=False}}{}
TODO: Doc
\index{\_\_init\_\_() (gfunc.data\_classes.GFuncNode method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.data_classes.GFuncNode.__init__}\pysiglinewithargsret{\bfcode{\_\_init\_\_}}{\emph{name}, \emph{species}, \emph{graph}, \emph{is\_target=False}, \emph{debug=False}}{}
TODO: Doc

\end{fulllineitems}

\index{get\_copy() (gfunc.data\_classes.GFuncNode method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.data_classes.GFuncNode.get_copy}\pysiglinewithargsret{\bfcode{get\_copy}}{}{}
Returns a deep copy of the node.

\end{fulllineitems}

\index{get\_sub\_scores() (gfunc.data\_classes.GFuncNode method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.data_classes.GFuncNode.get_sub_scores}\pysiglinewithargsret{\bfcode{get\_sub\_scores}}{\emph{target\_node}, \emph{graph}}{}
\end{fulllineitems}

\index{set\_data() (gfunc.data\_classes.GFuncNode method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.data_classes.GFuncNode.set_data}\pysiglinewithargsret{\bfcode{set\_data}}{\emph{data}, \emph{data\_type}}{}
TODO: Doc

\end{fulllineitems}

\index{total\_votes() (gfunc.data\_classes.GFuncNode method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.data_classes.GFuncNode.total_votes}\pysiglinewithargsret{\bfcode{total\_votes}}{}{}
\end{fulllineitems}


\end{fulllineitems}

\index{bunchify() (in module gfunc.data\_classes)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.data_classes.bunchify}\pysiglinewithargsret{\code{gfunc.data\_classes.}\bfcode{bunchify}}{\emph{dict\_tree}}{}
TODO: doc

\end{fulllineitems}

\phantomsection\label{code:module-gfunc.ensembl_data}\index{gfunc.ensembl\_data (module)}

\section{ensembl\_data.py}
\label{code:ensembl-data-py}
Code supporting automated retrieval/mirroring/processing of ensembl data files and directories.
\begin{description}
\item[{WARNING:}] \leavevmode
To some extent the functionality of this code is dependant on the
current formating of Ensembl's internal directory structure and how
it responts to \code{urllib2.urlopen(url)}.

\end{description}
\index{DataGrabber (class in gfunc.ensembl\_data)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.ensembl_data.DataGrabber}\pysiglinewithargsret{\strong{class }\code{gfunc.ensembl\_data.}\bfcode{DataGrabber}}{\emph{base\_url}, \emph{species}, \emph{data\_types}, \emph{base\_local\_path}, \emph{verbose=False}}{}
Class to manage conecting to ensembl-based ftp data dumps and retrieving them.
\index{\_\_init\_\_() (gfunc.ensembl\_data.DataGrabber method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.ensembl_data.DataGrabber.__init__}\pysiglinewithargsret{\bfcode{\_\_init\_\_}}{\emph{base\_url}, \emph{species}, \emph{data\_types}, \emph{base\_local\_path}, \emph{verbose=False}}{}
Initiate DataGrabber object for conecting to ensembl-based ftp data dumps.

\end{fulllineitems}

\index{settings() (gfunc.ensembl\_data.DataGrabber method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.ensembl_data.DataGrabber.settings}\pysiglinewithargsret{\bfcode{settings}}{}{}~\begin{description}
\item[{\emph{RETURNS:}}] \leavevmode\begin{itemize}
\item {} 
dict of current settings for ensembl access.

\end{itemize}

\end{description}

\end{fulllineitems}

\index{transfer\_data() (gfunc.ensembl\_data.DataGrabber method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.ensembl_data.DataGrabber.transfer_data}\pysiglinewithargsret{\bfcode{transfer\_data}}{\emph{unzip=False}}{}~\begin{description}
\item[{\emph{GIVEN:}}] \leavevmode\begin{itemize}
\item {} 
sufficently initiated \code{self} instance

\end{itemize}

\item[{\emph{DOES:}}] \leavevmode\begin{itemize}
\item {} 
{[}x{]} decides which method to use to get the data

\item {} 
{[}x{]} creates local directory if needed based on base\_local\_path

\item {} 
{[}x{]} initiates data transfer

\item {} 
{[}?{]} complains if it detects incomplete transfer

\item {} 
{[}?{]} if unzip evaluates to True, recursively unzip any files
with extentions suggesting they are compressed.

\end{itemize}

\item[{\emph{RETURNS:}}] \leavevmode\begin{itemize}
\item {} 
None

\end{itemize}

\end{description}

\end{fulllineitems}


\end{fulllineitems}

\index{check\_files() (in module gfunc.ensembl\_data)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.ensembl_data.check_files}\pysiglinewithargsret{\code{gfunc.ensembl\_data.}\bfcode{check\_files}}{\emph{dir\_path}, \emph{cksum\_data}}{}~\begin{description}
\item[{\emph{GIVEN:}}] \leavevmode\begin{itemize}
\item {} 
\code{dir\_path} = path to a directory containing files to be validated

\item {} 
\code{cksum\_data} = parsed contents of the CHECKSUMS file for this directory

\end{itemize}

\item[{\emph{DOES:}}] \leavevmode\begin{itemize}
\item {} 
Compares files in the directory with checksums in the CHECKSUMS file and scores them as PASS/FAIL

\item {} 
Documents and classifies discrepancies between files listed in CHECKSUMS file vs files actually
in the directory: classifies them as NOT\_IN\_CHECKSUMS or NOT\_IN\_DIR.

\end{itemize}

\item[{\emph{RETURNS:}}] \leavevmode\begin{itemize}
\item {} 
\code{results} = list of strings

\end{itemize}

\end{description}

\end{fulllineitems}

\index{validate\_downloads() (in module gfunc.ensembl\_data)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.ensembl_data.validate_downloads}\pysiglinewithargsret{\code{gfunc.ensembl\_data.}\bfcode{validate\_downloads}}{\emph{base\_dir}}{}~\begin{description}
\item[{\emph{GIVEN:}}] \leavevmode\begin{itemize}
\item {} 
\code{base\_dir} = top-level directory containing files to be validated.

\end{itemize}

\item[{\emph{DOES:}}] \leavevmode\begin{itemize}
\item {} 
Decends into \code{base\_dir} and records paths to all lower-level files.

\item {} 
Uses the CHECKSUMS files it finds to set up external system calls to \code{sum}
for each file listed in the direcory's CHECKSUMS file.

\item {} \begin{description}
\item[{A file named VALIDATIONS is created in each directory listing each file in the directory and one of the following outcomes: }] \leavevmode\begin{itemize}
\item {} 
PASS = passed checksum match

\item {} 
FAIL = failed checksum match

\item {} 
NOT\_IN\_CHECKSUMS = file found in directory but NOT listed in the CHECKSUMS file

\item {} 
NOT\_IN\_DIR = file listed in CHECKSUMS file, but not found in the directory.

\end{itemize}

\end{description}

\end{itemize}

\item[{\emph{RETURNS:}}] \leavevmode\begin{itemize}
\item {} 
\code{results} = a summary of all VALIDATIONS files (list of strings)

\end{itemize}

\end{description}

\end{fulllineitems}

\index{web\_ls() (in module gfunc.ensembl\_data)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.ensembl_data.web_ls}\pysiglinewithargsret{\code{gfunc.ensembl\_data.}\bfcode{web\_ls}}{\emph{url}}{}~\begin{description}
\item[{\emph{GIVEN:}}] \leavevmode\begin{itemize}
\item {} 
\code{url} = the url of an ensembl-based \href{ftp://}{ftp://} target

\end{itemize}

\item[{\emph{DOES:}}] \leavevmode\begin{itemize}
\item {} 
reads the url data and extracts file/directory names

\item {} 
stores info in a dict named \code{contents}; keyed by `dirs' or `files' and pointing to lists of respective urls.

\end{itemize}

\item[{\emph{RETURNS:}}] \leavevmode\begin{itemize}
\item {} 
\code{contents} dictionary.

\end{itemize}

\end{description}

\end{fulllineitems}

\index{web\_walk() (in module gfunc.ensembl\_data)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.ensembl_data.web_walk}\pysiglinewithargsret{\code{gfunc.ensembl\_data.}\bfcode{web\_walk}}{\emph{base\_url}}{}~\begin{description}
\item[{\emph{GIVEN:}}] \leavevmode\begin{itemize}
\item {} 
\code{base\_url} = the url of an ensembl-based \href{ftp://}{ftp://} target directory

\end{itemize}

\item[{\emph{DOES:}}] \leavevmode\begin{itemize}
\item {} 
recursively stores directories and file urls (uses \code{web\_ls}),
continues to follow directories until all file urls have been
collected below \code{base\_url}.

\end{itemize}

\item[{\emph{RETURNS:}}] \leavevmode\begin{itemize}
\item {} 
\code{file\_urls} = list of file url strings under \code{base\_url}.

\end{itemize}

\end{description}

\end{fulllineitems}

\phantomsection\label{code:module-gfunc.errors}\index{gfunc.errors (module)}

\section{errors.py}
\label{code:errors-py}
Code defining custom base error classes to provide a foundation for graceful error handling.
\index{GFuncError}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.errors.GFuncError}\pysigline{\strong{exception }\code{gfunc.errors.}\bfcode{GFuncError}}
Base class for exceptions in the gFunc package.

\end{fulllineitems}

\index{SystemCallError}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.errors.SystemCallError}\pysiglinewithargsret{\strong{exception }\code{gfunc.errors.}\bfcode{SystemCallError}}{\emph{errno}, \emph{strerror}, \emph{filename=None}}{}
Error raised when a problem occurs while attempting to run an external system call.
\begin{description}
\item[{Attributes:}] \leavevmode
\begin{DUlineblock}{0em}
\item[] \code{errno} -- return code from system call
\item[] \code{filename} -- file in volved if any
\item[] \code{strerror} -- error msg 
\end{DUlineblock}

\end{description}
\index{\_\_init\_\_() (gfunc.errors.SystemCallError method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.errors.SystemCallError.__init__}\pysiglinewithargsret{\bfcode{\_\_init\_\_}}{\emph{errno}, \emph{strerror}, \emph{filename=None}}{}
\end{fulllineitems}


\end{fulllineitems}

\index{UnsatisfiedDependencyError}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.errors.UnsatisfiedDependencyError}\pysigline{\strong{exception }\code{gfunc.errors.}\bfcode{UnsatisfiedDependencyError}}
Exception raised when gFunc can not find a suitable option to satisfy an external dependency.

\end{fulllineitems}

\phantomsection\label{code:module-gfunc.externals}\index{gfunc.externals (module)}

\section{externals.py}
\label{code:externals-py}
Code supporting running external system processes.
\index{mkdirp() (in module gfunc.externals)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.externals.mkdirp}\pysiglinewithargsret{\code{gfunc.externals.}\bfcode{mkdirp}}{\emph{path}}{}
Create new dir while creating any parent dirs in the path as needed.

\end{fulllineitems}

\index{runExternalApp() (in module gfunc.externals)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.externals.runExternalApp}\pysiglinewithargsret{\code{gfunc.externals.}\bfcode{runExternalApp}}{\emph{progName}, \emph{argStr}}{}
Convenience func to handle calling and monitoring output of external programs.

\end{fulllineitems}

\phantomsection\label{code:module-gfunc.fdr}\index{gfunc.fdr (module)}

\section{fdr.py}
\label{code:fdr-py}
Code supporting easy generation of emprical FDR for results.
\index{shuffle\_dict() (in module gfunc.fdr)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.fdr.shuffle_dict}\pysiglinewithargsret{\code{gfunc.fdr.}\bfcode{shuffle\_dict}}{\emph{original\_dict}, \emph{shuffle\_count}}{}~\begin{description}
\item[{\emph{GIVEN:}}] \leavevmode\begin{itemize}
\item {} 
\code{original\_dict}

\item {} 
\code{shuffle\_count}

\end{itemize}

\item[{\emph{RETURNS:}}] \leavevmode\begin{itemize}
\item {} 
generator that yields randomly shuffled key/value pairs as a new dict \code{shuffle\_count} times.

\end{itemize}

\end{description}

\end{fulllineitems}

\phantomsection\label{code:module-gfunc.fileIO}\index{gfunc.fileIO (module)}

\section{fileIO.py}
\label{code:fileio-py}
Code supporting reading and writing from files not related to specific parsers.
\index{tableFile2namedTuple() (in module gfunc.fileIO)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.fileIO.tableFile2namedTuple}\pysiglinewithargsret{\code{gfunc.fileIO.}\bfcode{tableFile2namedTuple}}{\emph{tablePath}, \emph{sep='\textbackslash{}t'}, \emph{headers=None}}{}
Returns namedTuple from table file using first row fields as
col headers or a list supplied by user.

\end{fulllineitems}

\index{walk\_dirs\_for\_fileName() (in module gfunc.fileIO)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.fileIO.walk_dirs_for_fileName}\pysiglinewithargsret{\code{gfunc.fileIO.}\bfcode{walk\_dirs\_for\_fileName}}{\emph{dir\_path}, \emph{pattern='*.xml'}}{}
Recursively collects file paths in a dir and subdirs.

\end{fulllineitems}

\phantomsection\label{code:module-gfunc.graphTools}\index{gfunc.graphTools (module)}

\section{graphTools.py}
\label{code:graphtools-py}
Code supporting building and querying the graphs.
\index{GraphBuilder (class in gfunc.graphTools)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.graphTools.GraphBuilder}\pysiglinewithargsret{\strong{class }\code{gfunc.graphTools.}\bfcode{GraphBuilder}}{\emph{parsers}}{}~\index{\_\_init\_\_() (gfunc.graphTools.GraphBuilder method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.graphTools.GraphBuilder.__init__}\pysiglinewithargsret{\bfcode{\_\_init\_\_}}{\emph{parsers}}{}
\end{fulllineitems}

\index{map\_registries\_to\_graph() (gfunc.graphTools.GraphBuilder method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.graphTools.GraphBuilder.map_registries_to_graph}\pysiglinewithargsret{\bfcode{map\_registries\_to\_graph}}{\emph{nodes=True}, \emph{edges=True}}{}
Iterates through each registry creating graph nodes and edges.  Returns GraphHandler.

\end{fulllineitems}

\index{populate\_registries() (gfunc.graphTools.GraphBuilder method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.graphTools.GraphBuilder.populate_registries}\pysiglinewithargsret{\bfcode{populate\_registries}}{}{}
Iterates through provided parsers and calls the parsers `resgister\_nodes\_and\_edges'
method to populate/update the relevant registries (node/edge\_dict).

\end{fulllineitems}


\end{fulllineitems}

\index{GraphHandler (class in gfunc.graphTools)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.graphTools.GraphHandler}\pysiglinewithargsret{\strong{class }\code{gfunc.graphTools.}\bfcode{GraphHandler}}{\emph{node\_dict}, \emph{edge\_dict}, \emph{graph}}{}
TODO: Doc
\index{\_\_init\_\_() (gfunc.graphTools.GraphHandler method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.graphTools.GraphHandler.__init__}\pysiglinewithargsret{\bfcode{\_\_init\_\_}}{\emph{node\_dict}, \emph{edge\_dict}, \emph{graph}}{}
TODO: Doc

\end{fulllineitems}

\index{clone\_node\_as\_target() (gfunc.graphTools.GraphHandler method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.graphTools.GraphHandler.clone_node_as_target}\pysiglinewithargsret{\bfcode{clone\_node\_as\_target}}{\emph{node\_name}}{}
TODO: Doc

\end{fulllineitems}

\index{install\_metric\_handlers() (gfunc.graphTools.GraphHandler method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.graphTools.GraphHandler.install_metric_handlers}\pysiglinewithargsret{\bfcode{install\_metric\_handlers}}{\emph{rel\_hndler}, \emph{vote\_hndlr}}{}
TODO: Doc

\end{fulllineitems}

\index{install\_target() (gfunc.graphTools.GraphHandler method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.graphTools.GraphHandler.install_target}\pysiglinewithargsret{\bfcode{install\_target}}{}{}
TODO: Doc

\end{fulllineitems}

\index{measure\_relations() (gfunc.graphTools.GraphHandler method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.graphTools.GraphHandler.measure_relations}\pysiglinewithargsret{\bfcode{measure\_relations}}{}{}
Cues RelationsHandler to do its thing after pasing it self.edge\_dict.

\end{fulllineitems}

\index{take\_votes() (gfunc.graphTools.GraphHandler method)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.graphTools.GraphHandler.take_votes}\pysiglinewithargsret{\bfcode{take\_votes}}{\emph{node\_list}, \emph{poll\_func=None}}{}
Cues VoteHandler to do its thing after pasing it a list of specific
GFuncNode objects.

Example:
\textgreater{}\textgreater{}\textgreater{} node\_list = {[}node for node in node\_dict.itervalues() if node.species == `Anopheles gambie'{]}.

\end{fulllineitems}


\end{fulllineitems}

\phantomsection\label{code:module-gfunc.maths}\index{gfunc.maths (module)}

\section{maths.py}
\label{code:maths-py}
Code supporting specialized calculations for gfunc.
\index{bayesian\_score() (in module gfunc.maths)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.maths.bayesian_score}\pysiglinewithargsret{\code{gfunc.maths.}\bfcode{bayesian\_score}}{\emph{c}, \emph{m}, \emph{n}, \emph{scores}, \emph{scale\_mod=1}}{}~
\begin{DUlineblock}{0em}
\item[] BS = ((c * m) + sum({[}x for x in scores{]})) / (n + c)
\item[] 
\item[] Where:
\item[] 
\item[] \code{n}: number of votes for THIS item
\item[] \code{C}: median number of votes for all items that got at least 1 vote (weighting or dampening factor)
\item[] \code{m}: median UNweighted score for all items that got at least 1 vote  
\end{DUlineblock}

\end{fulllineitems}

\index{weight\_d\_for\_ptci() (in module gfunc.maths)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.maths.weight_d_for_ptci}\pysiglinewithargsret{\code{gfunc.maths.}\bfcode{weight\_d\_for\_ptci}}{\emph{d\_i}, \emph{d\_min}, \emph{d\_max}, \emph{w\_min=1.0}, \emph{w\_max=1.1}}{}
Scaling function to transform `d' onto a weight-spectrum to either
punish or reward the final ptci score based on the phylogenetic distance
between the two current species as it relates to the range of phylogenetic
distances in the data set.

Default weight scale is no change for the shortest distance (return 1.0) to
a 10\% reward for the longest distance (return 1.1).

\end{fulllineitems}

\phantomsection\label{code:module-gfunc.motifs}\index{gfunc.motifs (module)}

\section{motifs.py}
\label{code:motifs-py}
Code supporting the searching, recording, and analysis of sequence motifs for gFunc.
\index{load\_MOODS\_result() (in module gfunc.motifs)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.motifs.load_MOODS_result}\pysiglinewithargsret{\code{gfunc.motifs.}\bfcode{load\_MOODS\_result}}{\emph{in\_path}}{}~\begin{description}
\item[{\emph{GIVEN:}}] \leavevmode\begin{itemize}
\item {} 
\code{in\_path}: path to store results

\end{itemize}

\item[{\emph{DOES:}}] \leavevmode\begin{itemize}
\item {} 
loads \code{processed\_moods\_result\_dict} from binary pickle to \code{in\_path}.

\end{itemize}

\item[{\emph{RETURNS:}}] \leavevmode\begin{itemize}
\item {} 
un-pickled MOODS result object.

\end{itemize}

\end{description}

\end{fulllineitems}

\index{motif\_profiles\_weighted\_by\_score() (in module gfunc.motifs)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.motifs.motif_profiles_weighted_by_score}\pysiglinewithargsret{\code{gfunc.motifs.}\bfcode{motif\_profiles\_weighted\_by\_score}}{\emph{processed\_moods\_result\_dict}}{}~\begin{description}
\item[{\emph{GIVEN:}}] \leavevmode\begin{itemize}
\item {} 
processed\_moods\_result\_dict: output from (def \code{process\_MOODS\_results()})

\end{itemize}

\item[{\emph{DOES:}}] \leavevmode\begin{itemize}
\item {} 
iterates through \code{processed\_moods\_result\_dict} and calculates a motif presence score (mps)
for each motifName:seqName pair by summing the all positive \code{site\_scores} for motifName in seqName.

\item {} 
mps are recorded in a \code{pandas.DataFrame} with columns = motifName and indexs = seqName.

\end{itemize}

\item[{\emph{RETURNS:}}] \leavevmode\begin{itemize}
\item {} 
mps\_table: \code{pandas.DataFrame} constructed as above.

\end{itemize}

\end{description}

\end{fulllineitems}

\index{process\_MOODS\_results() (in module gfunc.motifs)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.motifs.process_MOODS_results}\pysiglinewithargsret{\code{gfunc.motifs.}\bfcode{process\_MOODS\_results}}{\emph{moods\_result\_dict}, \emph{motif\_names}}{}~\begin{description}
\item[{\emph{GIVEN:}}] \leavevmode\begin{itemize}
\item {} 
\code{moods\_result\_dict}: orderedDict ???is this true??? of moods\_results\_tuples keyed by seqName/geneName

\item {} 
\code{motif\_names}: correctly ordered motif names (Motifs.motifs.keys())

\end{itemize}

\item[{\emph{DOES:}}] \leavevmode\begin{itemize}
\item {} 
converts \code{moods\_result\_dict} into a three key'd multi-level dict as follows:
\code{processed\_moods\_result\_dict{[}SeqName{]}{[}motifName{]}{[}location{]} = score}

\end{itemize}

\item[{\emph{RETURNS:}}] \leavevmode\begin{itemize}
\item {} 
\code{processed\_moods\_result\_dict}

\end{itemize}

\end{description}

\end{fulllineitems}

\index{save\_MOODS\_result() (in module gfunc.motifs)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.motifs.save_MOODS_result}\pysiglinewithargsret{\code{gfunc.motifs.}\bfcode{save\_MOODS\_result}}{\emph{moods\_hits}, \emph{out\_path}}{}~\begin{description}
\item[{\emph{GIVEN:}}] \leavevmode\begin{itemize}
\item {} 
\code{moods\_hits}: non-processed result from \code{Motif.scan\_seq()} or \code{Motif.scan\_seqDict()}.

\item {} 
\code{out\_path}: path to store results

\end{itemize}

\item[{\emph{DOES:}}] \leavevmode\begin{itemize}
\item {} 
stores \code{moods\_hits} as binary pickle to \code{out\_path}.

\end{itemize}

\item[{\emph{RETURNS:}}] \leavevmode\begin{itemize}
\item {} 
\code{None}

\end{itemize}

\end{description}

\end{fulllineitems}

\phantomsection\label{code:module-gfunc.stats}\index{gfunc.stats (module)}

\section{stats.py}
\label{code:stats-py}
Code supporting calculations of statistical probabilities for gfunc.
\index{basic\_bootstrap\_est() (in module gfunc.stats)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.stats.basic_bootstrap_est}\pysiglinewithargsret{\code{gfunc.stats.}\bfcode{basic\_bootstrap\_est}}{\emph{vec}, \emph{reps=1000}}{}~\begin{description}
\item[{\emph{GIVEN:}}] \leavevmode\begin{itemize}
\item {} 
\code{vec} = vector of sample values

\item {} 
\code{reps} = number of resampling reps

\end{itemize}

\item[{\emph{DOES:}}] \leavevmode\begin{itemize}
\item {} 
Resample w/ replacement \code{reps} times and record the medians

\item {} 
Calculate stdv of resampled medians which should approach
the actual SE as \code{reps} approaches \code{inf}.

\item {} 
Calculate the 95\% CI bounds.

\end{itemize}

\item[{\emph{RETURNS:}}] \leavevmode\begin{itemize}
\item {} 
tuple({[}\emph{median of resampled medians}, \emph{SE est}, \emph{loBound}, \emph{hiBound}{]})

\end{itemize}

\end{description}

\end{fulllineitems}

\index{benjHochFDR() (in module gfunc.stats)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.stats.benjHochFDR}\pysiglinewithargsret{\code{gfunc.stats.}\bfcode{benjHochFDR}}{\emph{table}, \emph{pValColumn=-1}}{}~\begin{description}
\item[{\emph{GIVEN:}}] \leavevmode\begin{itemize}
\item {} 
\code{table}: 2D list(\emph{hypothesis},*p-value*) hypothesis could = \emph{geneName} tested for enrichment

\item {} 
\code{pValColumn}: integer of column index containing the \emph{p-value}.

\end{itemize}

\item[{\emph{DOES:}}] \leavevmode\begin{itemize}
\item {} 
Calculates the Benjamini-Hochberg adjusted \emph{p-values}

\end{itemize}

\item[{\emph{RETURNS:}}] \leavevmode\begin{itemize}
\item {} 
a new version of \code{table} with an extra column added to the end representing the BH corrected \emph{p-values}

\end{itemize}

\end{description}

\end{fulllineitems}

\index{binComb() (in module gfunc.stats)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.stats.binComb}\pysiglinewithargsret{\code{gfunc.stats.}\bfcode{binComb}}{\emph{n}, \emph{k}}{}~\begin{description}
\item[{\emph{GIVEN:}}] \leavevmode\begin{itemize}
\item {} 
\code{n}

\item {} 
\code{k}

\end{itemize}

\item[{\emph{DOES:}}] \leavevmode\begin{itemize}
\item {} 
Computes \code{n} \emph{choose} \code{k}.

\end{itemize}

\item[{\emph{RETURNS:}}] \leavevmode\begin{itemize}
\item {} 
The number of ways \code{k} objects can be sampled from a population of size \code{n}.

\end{itemize}

\end{description}

\end{fulllineitems}

\index{binomialPval() (in module gfunc.stats)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.stats.binomialPval}\pysiglinewithargsret{\code{gfunc.stats.}\bfcode{binomialPval}}{\emph{n}, \emph{k}, \emph{p}}{}~\begin{description}
\item[{\emph{RETURNS:} }] \leavevmode\begin{itemize}
\item {} 
exact binomial P-value.

\end{itemize}

\end{description}

\begin{DUlineblock}{0em}
\item[] \code{n} = number of trials
\item[] \code{k} = number of successes
\item[] \code{p} = probability of a success
\item[] 
\item[] \emph{P(k succeses in n trials) = choose(n,k) (p\textasciicircum{}k) ((1-p)\textasciicircum{}(n-k))}
\end{DUlineblock}

\end{fulllineitems}

\index{binomialPval\_gte() (in module gfunc.stats)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.stats.binomialPval_gte}\pysiglinewithargsret{\code{gfunc.stats.}\bfcode{binomialPval\_gte}}{\emph{n}, \emph{k}, \emph{p}}{}~\begin{description}
\item[{\emph{RETURNS:} }] \leavevmode\begin{itemize}
\item {} 
binomial \emph{p-value} of \code{k} or greater successes in \code{n} trials with probability of success for each trial \code{p}.

\end{itemize}

\end{description}

\begin{DUlineblock}{0em}
\item[] \code{n} = number of trials
\item[] \code{k} = number of successes
\item[] \code{p} = probability of a success
\item[] 
\item[] \emph{sum( choose(n,k) (p\textasciicircum{}k) ( (1-p)\textasciicircum{}(n-k) ) )} as \code{k} goes from \code{k} to \code{n}
\end{DUlineblock}

\end{fulllineitems}

\index{cumHypergeoP() (in module gfunc.stats)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.stats.cumHypergeoP}\pysiglinewithargsret{\code{gfunc.stats.}\bfcode{cumHypergeoP}}{\emph{n}, \emph{i}, \emph{m}, \emph{N}}{}~
\begin{DUlineblock}{0em}
\item[] Calculates the cumulative hypergeometric \emph{p-value} for variables:
\item[] 
\item[] \code{n} = number of positives in population
\item[] \code{i} = number of positives in sample
\item[] \code{m} = number of negatives in population
\item[] \code{N} = sample size
\item[] 
\item[] \emph{P(i) = sum({[}as i-\textgreater{}N{]} (choose(n,i)choose(m,N-i))/choose(n+m,N))}
\item[] 
\item[] For more details -\textgreater{} \href{http://mathworld.wolfram.com/HypergeometricDistribution.html}{http://mathworld.wolfram.com/HypergeometricDistribution.html}
\end{DUlineblock}

\end{fulllineitems}

\index{hypergeoP() (in module gfunc.stats)}

\begin{fulllineitems}
\phantomsection\label{code:gfunc.stats.hypergeoP}\pysiglinewithargsret{\code{gfunc.stats.}\bfcode{hypergeoP}}{\emph{n}, \emph{i}, \emph{m}, \emph{N}}{}~
\begin{DUlineblock}{0em}
\item[] Calculates the non-cumulative hypergeometric \emph{p-value} for variables:
\item[] 
\item[] \code{n} = number of positives in population
\item[] \code{i} = number of positives in sample
\item[] \code{m} = number of negatives in population
\item[] \code{N} = sample size
\item[] 
\item[] \emph{P(x=i) = (choose(n,i)choose(m,N-i))/choose(n+m,N)}
\item[] 
\item[] For more details -\textgreater{} \href{http://mathworld.wolfram.com/HypergeometricDistribution.html}{http://mathworld.wolfram.com/HypergeometricDistribution.html}
\end{DUlineblock}

\end{fulllineitems}

\phantomsection\label{code:module-gfunc.xpermutations}\index{gfunc.xpermutations (module)}\phantomsection\label{code:module-gfunc.galaxy_tools}\index{gfunc.galaxy\_tools (module)}

\section{galaxy\_tools}
\label{code:galaxy-tools}
Code supporting the integration of gFunc with the Galaxy paradigm.

So far this is empty!


\chapter{Indices and tables}
\label{index:indices-and-tables}\begin{itemize}
\item {} 
\emph{genindex}

\item {} 
\emph{modindex}

\item {} 
\emph{search}

\end{itemize}


\renewcommand{\indexname}{Python Module Index}
\begin{theindex}
\def\bigletter#1{{\Large\sffamily#1}\nopagebreak\vspace{1mm}}
\bigletter{g}
\item {\texttt{gfunc.\_\_init\_\_}}, \pageref{code:module-gfunc.__init__}
\item {\texttt{gfunc.analysis\_classes}}, \pageref{code:module-gfunc.analysis_classes}
\item {\texttt{gfunc.clustering}}, \pageref{code:module-gfunc.clustering}
\item {\texttt{gfunc.data\_classes}}, \pageref{code:module-gfunc.data_classes}
\item {\texttt{gfunc.ensembl\_data}}, \pageref{code:module-gfunc.ensembl_data}
\item {\texttt{gfunc.errors}}, \pageref{code:module-gfunc.errors}
\item {\texttt{gfunc.externals}}, \pageref{code:module-gfunc.externals}
\item {\texttt{gfunc.fdr}}, \pageref{code:module-gfunc.fdr}
\item {\texttt{gfunc.fileIO}}, \pageref{code:module-gfunc.fileIO}
\item {\texttt{gfunc.galaxy\_tools}}, \pageref{code:module-gfunc.galaxy_tools}
\item {\texttt{gfunc.graphTools}}, \pageref{code:module-gfunc.graphTools}
\item {\texttt{gfunc.maths}}, \pageref{code:module-gfunc.maths}
\item {\texttt{gfunc.motifs}}, \pageref{code:module-gfunc.motifs}
\item {\texttt{gfunc.parsers.\_\_init\_\_}}, \pageref{code:module-gfunc.parsers.__init__}
\item {\texttt{gfunc.parsers.base}}, \pageref{code:module-gfunc.parsers.base}
\item {\texttt{gfunc.parsers.Cufflinks}}, \pageref{code:module-gfunc.parsers.Cufflinks}
\item {\texttt{gfunc.parsers.edge\_lists}}, \pageref{code:module-gfunc.parsers.edge_lists}
\item {\texttt{gfunc.parsers.ETE}}, \pageref{code:module-gfunc.parsers.ETE}
\item {\texttt{gfunc.parsers.GTF}}, \pageref{code:module-gfunc.parsers.GTF}
\item {\texttt{gfunc.parsers.JASPAR}}, \pageref{code:module-gfunc.parsers.JASPAR}
\item {\texttt{gfunc.parsers.MAST}}, \pageref{code:module-gfunc.parsers.MAST}
\item {\texttt{gfunc.stats}}, \pageref{code:module-gfunc.stats}
\item {\texttt{gfunc.xpermutations}}, \pageref{code:module-gfunc.xpermutations}
\end{theindex}

\renewcommand{\indexname}{Index}
\printindex
\end{document}
